generate YAML manifests for NSTemplateTier resources

xcoulon opened the pull request:
Use Go templates to embed the Namespace templates and produce the OpenShift Template manifests.
Also, the new `NSTemplateTierGenerator` type loads all metadata once, so there's no need to parse the `metadata.yaml` file each time a set of NSTemplateTier manifests must be generated.

For testing: 
- verifies that the actual output (`[]byte`) matches the expectation and that it's correctly formatted by trying to parse it and convert it to "OpenShift" objects.
- use a specific a set of templates in `test/templates/nstemplatetiers` along with an arbitrary `metadata.yaml` file (ie, with different hashes for each file), so we don't depend on templates aimed at being used in production, nor do we need to keep track of their commit hashes.
- make sure that all 3 pkgs (deploy/run/test) are consistent in their naming: `deploy/templates/nstemplatetiers`, `pkg/templates/nstemplatetiers` and `test/templates/nstemplatetiers`.
- in addition, `make generate` uses `go-bindata` to also generate the `Asset()` func in `test/templates/nstemplatetiers/nstemplatetier_assets.go` (similarly to `pkg/templates/nstemplatetiers/nstemplatetier_assets.go` for the real/production templates)
the `GenerateNSTemplateTiers` function has a pointer receiver which is initialized with the `Asset` function to use, so we can easily switch to the test templates in the tests


Fixes CRT-258

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

alexeykazakov commented on pkg/templates/nstemplatetier_generate_test.go:
```diff
@@ -0,0 +1,103 @@
+		result, err := templates.GenerateNSTemplateTierManifest(tier)
+		// then
+		require.NoError(t, err)
+		expected := `kind: NSTemplateTier
```
So, every time we change any of the templates we will need to update this expected string too. It can be pretty inconvenient. If we use some util to re-generate the template then we may compromise the test.

What if we:
- construct NSTemplateTier structs (https://github.com/codeready-toolchain/api/blob/master/pkg/apis/toolchain/v1alpha1/nstemplatetier_types.go#L31) for expected tiers (all of them) with all not-optional fields set up.
- unmarshal all the generated tier yaml manifests into NSTemplateTier structs
- smart assert equality of these expected and actual tiers.

Expected revisions and template values could be loaded from Asset.

Regarding revision values. The revisions (commit SHAs) will be actually different. They are auto-generated in build time. So they will be different in the binaries built from master and from the PR branch. So, we can't hardcode them in the tests. Unless I'm missing something.
We will need to load them from Asset and check if they match.

Also, in our API (https://github.com/codeready-toolchain/api/blob/master///pkg/apis/toolchain/v1alpha1/nstemplateset_types.go#L33) templates are currently optional strings and are supposed to be used for custom templates only. But it seems that you actually are going to pass the templates in the tier manifest as yaml and it's not optional (so it's not for custom templates anymore).
We need to align our API with that approach.
We will probably need to change the Template field of the Namespace struct from `string` to https://github.com/openshift/api/blob/master/template/v1/types.go#L15 and make it required.

xcoulon replied:
@alexeykazakov I'm going to see how to address your concerns with regards to tests stability as the templates and commit hash evolve with time.

xcoulon commented on pkg/templates/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,151 @@
+// ------
+func GenerateNSTemplateTierManifest(tier string) ([]byte, error) {
+	if nstemplatetierTmpl == nil {
+		return nil, errors.Errorf("Cannot generate the '%s' NSTemplateTier manifest", tier)
```
ok, I can do that

xcoulon commented on pkg/templates/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,151 @@
+	return parseRevisions(metadata, tier)
+}
+
+// GetRevisions return a map of revisions indexed by their associated template kind
```
oh really? 😬

xcoulon commented:
@alexeykazakov I've addressed your concerns about testing stability while the deployed templates will evolve over time with the following actions:

- created a set of templates in `test/templates/nstemplatetiers`, along with an arbitrary `metadata.yaml` file (ie, with different hashes for each file)
- moved the code from `pkg/templates` to `pkg/templates/nstemplatetiers` so all 3 pkgs are consistent: `deploy/templates/nstemplatetiers`, `pkg/templates/nstemplatetiers` and `test/templates/nstemplatetiers`. 
- in addition, `make generate` uses `go-bindata` to generate the `Asset()` func in `test/templates/nstemplatetiers/nstemplatetier_assets.go` (similarly to `pkg/templates/nstemplatetiers/nstemplatetier_assets.go` for the real/production templates)
- the `GenerateNSTemplateTiers` function has now a pointer receiver which is initialized with the `Asset` function to use, so we can easily switch to the test templates in the tests
- also, this `NSTemplateTierGenerator` type loads all metadata once, so there's no need to parse the `metadata.yaml` file each time a set of NSTemplateTier manifests must be generated.

Please let me know what you think about all this 😉

xcoulon commented on deploy/templates/nstemplatetiers/metadata.yaml:
```diff
@@ -0,0 +1,6 @@
+advanced-code: e3f140c
```
right, let me remove it.

xcoulon replied:
removed in f256204

xcoulon commented on .gitignore:
```diff
@@ -8,8 +8,9 @@ bin/
+pkg/templates/nstemplatetiers/nstemplatetier_assets.go
+test/templates/nstemplatetiers/nstemplatetier_assets.go
+deploy/templates/nstemplatetiers/metadata.yaml
 
```
I'm sorry but here my repo is has no unstaged/uncommitted changes after running `make clean && make build`

MatousJobanek replied:
@alexeykazakov the reason is that these files were moved to a new location (folder) and the gitignore was changed too

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,187 @@
+
+// NSTemplateTierGenerator the NSTemplateTier manifest generator
+type NSTemplateTierGenerator struct {
+	asset     func(name string) ([]byte, error) // the func which gives access to the
```
good catch! It was a left-over from some refactoring, and now removed in dda4b75

xcoulon replied:
... and restored in 74d08a5 after checking the comment below :)

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,187 @@
+	sort.Strings(nsTypes)
+	for _, nsType := range nsTypes {
+		// get the content of the `-<t>.yaml` file
+		tmpl, err := Asset(fmt.Sprintf("%s-%s.yaml", tier, nsType))
```
@alexeykazakov good catch! yes, I already used the `asset` on line 46, but I missed that one since for now both `deploy/` and `test/` pkgs have the same namespace templates.

xcoulon replied:
fixed in 74d08a5

xcoulon commented:
@alexeykazakov I addressed all your concerns. Could you please give this PR a second look?

xcoulon commented on make/clean.mk:
```diff
@@ -2,5 +2,5 @@
-	$(Q)-rm pkg/templates/template_contents.go
-	$(Q)-rm deploy/nstemplatetiers/metadata.yaml
+	$(Q)-rm pkg/templates/nstemplatetiers/nstemplatetier_assets.go 
+	$(Q)-rm deploy/templates/nstemplatetiers/metadata.yaml
```
I've tried this, but I still got the error displayed in the terminal, with a `(ignored)` comment when the files didn't exist:
```
rm -rf  ./build/_output ./vendor
go clean  ./...
rm deploy/templates/nstemplatetiers/metadata.yaml
rm: deploy/templates/nstemplatetiers/metadata.yaml: No such file or directory
make: [clean] Error 1 (ignored)
rm pkg/templates/nstemplatetiers/nstemplatetier_assets.go 2>/dev/null
make: [clean] Error 1 (ignored)
rm test/templates/nstemplatetiers/nstemplatetier_assets.go 2>/dev/null
make: [clean] Error 1 (ignored)
```

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,209 @@
+			continue
+		}
+		if _, ok := revisions[tierKind]; !ok {
+			revisions[tierKind] = make(map[string]string, 3) // expect 3 entries: 'code', 'dev' and 'stage'
```
that's the initial capacity. It's not going to be an issue if we have more entries

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,209 @@
+    template:
+{{ .Template }}{{ end }}`)
+	if initTmplErr != nil {
+		log.Error(initTmplErr, "failed to initialize NSTemplateTier template")
```
we only need to load it once, hence the reason for doing it in `init()`

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,209 @@
+		// create a new entry if needed
+		revision, ok := revision.(string)
+		if !ok {
+			log.Info("invalid namespace template filename revision. Expected a string", "filename", filename, "revision", revision)
```
ok, I will return an error, then

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,209 @@
+	for filename, revision := range data {
+		data := strings.Split(filename, "-")
+		if len(data) != 2 {
+			log.Info("invalid namespace template filename. Expected format: '<tier_kind>-<namespace_kind>'", "filename", filename)
```
ok, I will return an error here too, then

MatousJobanek replied:
didn't you forget to do that? same also for this one https://github.com/codeready-toolchain/host-operator/pull/69/files#r329525757

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator_test.go:
```diff
@@ -0,0 +1,140 @@
+		result, err := g.GenerateManifest(tier)
+		// then
+		require.NoError(t, err)
+		expected := `kind: NSTemplateTier
```
sure

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,209 @@
+//   - code: <[]byte>
+//   - dev: <[]byte>
+//   - stage: <[]byte>
+func (g NSTemplateTierGenerator) GenerateAllManifests() (map[string][]byte, error) {
```
thanks for your feedback. Indeed, it's going to be more straightforward if we directly build the objects instead of generating a large YAML file here, and parse it in a subsequent PR. So let me refactor this PR to adapt the code to your suggestions.

alexeykazakov commented on pkg/templates/nstemplatetiers/nstemplatetier_generator_test.go:
```diff
@@ -0,0 +1,140 @@
+	})
+}
+
+// generates uses a golang template which is execute and whose output is decoded into an toolchainv1alpha1.NSTemplateTier.
```
```suggestion
// newNSTemplateTierFromYAML generates toolchainv1alpha1.NSTemplateTier using a golang template which is applied to the given tier.
```

alexeykazakov commented:
BTW, this one: can be easily tested - https://codecov.io/gh/codeready-toolchain/host-operator/pull/69/diff#D22-138
right?

xcoulon commented:
> BTW, this one can be easily tested - https://codecov.io/gh/codeready-toolchain/host-operator/pull/69/diff#D22-138
> right?

yes, let me add a test for that case

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -143,20 +132,22 @@ func (g NSTemplateTierGenerator) GenerateAllManifests() (map[string][]byte, erro
-		Template string
+	obj := toolchainv1alpha1.NSTemplateTier{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "toolchain.dev.openshift.com/v1alpha1",
```
@alexeykazakov I needed to add it because the `expected` object which is generated from a YAML source has the info. Otherwise, the comparison would fail :/

alexeykazakov replied:
Hm... then maybe let's remove it from the test yamls too? :)
We are really not supposed to set group and kind in objets. As far as I understand it's up to the go client to map the go type (`toolchainv1alpha1.NSTemplateTier` in this case) to the registered GVK.
In best case this meta will be ignored (so it doesn't give us anything if we compare it in the tests). In worst case it can cause issues if we forget to update these hardcoded string when switching to a newer API version (so GVK will be different).

xcoulon replied:
what I can do, though, is compare the `ObjectMeta` and `Spec` portions of the `NSTemplateTier` objects, ignoring the `TypeMeta` (assuming it will be valid)

alexeykazakov replied:
> what I can do, though, is compare the ObjectMeta and Spec portions of the NSTemplateTier objects, ignoring the TypeMeta (assuming it will be valid): see de8f6e5

I prefer this option ^ since it doesn't require unnecessary code in production. So, it's isolated in the tests.

xcoulon replied:
ok, so let me revert 83a5965, then

alexeykazakov replied:
We still need to remove the entire `TypeMeta` :)

xcoulon replied:
🤦‍♂ (and done in 1460ce6)

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -106,19 +88,26 @@ func parseAllRevisions(metadata []byte) (map[string]map[string]string, error) {
-		manifests[tierKind], err = g.GenerateManifest(tierKind)
+func (g NSTemplateTierGenerator) GenerateAllManifests(namespace string) ([]toolchainv1alpha1.NSTemplateTier, error) {
+	objects := make([]toolchainv1alpha1.NSTemplateTier, 0, len(g.revisions))
+	// retrieve the tier names and order them, so we can order them, and compare
```
ahem... yes, indeed!

xcoulon replied:
> > BTW, this one can be easily tested - https://codecov.io/gh/codeready-toolchain/host-operator/pull/69/diff#D22-138
> > right?
> 
> yes, let me add a test for that case

done in 6baa8cf

xcoulon replied:
done in ccf2687

xcoulon commented:
> > BTW, this one can be easily tested - https://codecov.io/gh/codeready-toolchain/host-operator/pull/69/diff#D22-138
> > right?
> 
> yes, let me add a test for that case

done in done in 6baa8cf

MatousJobanek commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,201 @@
+func NewNSTemplateTierGenerator(s *runtime.Scheme, asset func(name string) ([]byte, error)) (*NSTemplateTierGenerator, error) {
+	metadata, err := asset("metadata.yaml")
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize the NSTemplateTierGenerator")
```
```suggestion
		return nil, errors.Wrapf(err, "unable to get the metadata.yaml asset while initializing NSTemplateTierGenerator")
```

xcoulon replied:
ok, this change may make sense, although, I believe that the `Asset` func would return an error which mentions the `metadata.yaml` file

MatousJobanek commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,201 @@
+	}
+	revisions, err := parseAllRevisions(metadata)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize the NSTemplateTierGenerator")
```
```suggestion
		return nil, errors.Wrapf(err, "unable to parse revisions while initializing NSTemplateTierGenerator")
```

xcoulon replied:
Same here: `parseAllRevisions()` already returns a contextual error, so I think it's good enough to wrap it with what is being done in the `NewNSTemplateTierGenerator`: initializing an NSTemplateTierGenerator. Besides, the error message will be something like: `unable to initialize the NSTemplateTierGenerator: unable to parse all template revisions: <underlying error>". I believe it's not necessary to give more info about the func that was called, but only to give info about the current function.

MatousJobanek commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,201 @@
+	for tier := range g.revisions {
+		tiers = append(tiers, tier)
+	}
+	sort.Strings(tiers)
```
I guess that he update is not implemented yet (which is fine) but I really don't see the purpose of this sorting. We should go through the tiers one by one and compare the particular newly generated  `NSTemplateTier` with the one existing in the cluster and update the particular one if needed. I think that there is no need for keeping the order.

xcoulon replied:
the sorting here helped with the comparison in the test, but I guess I could return a map instead of a slice, so there would be no need for such sorting. Let me do that

xcoulon replied:
done in d68ff4e

MatousJobanek commented on pkg/templates/nstemplatetiers/nstemplatetier_generator.go:
```diff
@@ -0,0 +1,201 @@
+	for nsType := range revisions {
+		nsTypes = append(nsTypes, nsType)
+	}
+	sort.Strings(nsTypes)
```
on the other hand, I understand the purpose of this sorting. In this case you really care about the order in case you want to use `reflect.DeepEqual` function. 
another option would be implementing our own equal function for the `NSTemplateTierSpec` as is done for `NSTemplateSetSpec` https://github.com/codeready-toolchain/api/blob/master/pkg/apis/toolchain/v1alpha1/nstemplateset_types.go#L78
both options are fine, so it's up to you.

xcoulon replied:
good point, let me do that instead.

alexeykazakov replied:
I still think it could be useful to have Namespaces sorted in NSTemplateTier.Spec. So we can guarantee that we will get the same yaml (if we marshal the struct) every time we generate a tier for the same templates. It can be useful in prod if we put templates to annotations in the future.

xcoulon commented:
@MatousJobanek I've replaced the `log.Info` with a `return nil, err..` when the filename or the hash is invalid in 76f159f (but I can't find the original request in the comments, hence posting it here)

MatousJobanek commented on pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go:
```diff
@@ -0,0 +1,114 @@
+			assert.NotEmpty(t, g.revisions[tier])
+		}
+	})
+
```
may we have also a test that calls `g.NewNSTemplateTiers(namespace)` for production files? For now, there is no way we could verify that all prod yaml files have correct formatting so it won't fail while deploying to cluster, right? I know that it should be covered by e2e tests, but it would be nice if we could catch it already in a unit test. Could something like this work?: 
```go
t.Run("tiers", func(t *testing.T) {
	// given
	g, err := NewNSTemplateTierGenerator(s, Asset)
	require.NoError(t, err)

	// when
	tiers, err := g.NewNSTemplateTiers("my-cool-ns")
	require.NoError(t, err)
	require.NotEmpty(t, tiers)

	for _, actual := range tiers {
		tierName := actual.Name
		assert.Equal(t, "my-cool-ns", actual.Namespace)
		assert.Len(t, g.revisions[tierName], len(actual.Spec.Namespaces))
		for nsType, rev := range g.revisions[tierName] {
			found := false
			for _, ns := range actual.Spec.Namespaces {
				if ns.Type == nsType {
					found = true
					assert.Equal(t, rev, ns.Revision)
					asset, err := Asset(fmt.Sprintf("%s-%s.yaml", tierName, nsType))
					require.NoError(t, err)
					template, err := decodeTemplate(serializer.NewCodecFactory(s).UniversalDeserializer(), asset)
					require.NoError(t, err)
					assert.Equal(t, *template, ns.Template)
					break
				}
			}
			assert.True(t, found, "the namespace with the type wasn't found", "ns-type", nsType)
		}
	}
})
```

xcoulon replied:
ok, let me add something like that

xcoulon replied:
@MatousJobanek after giving it some thoughts, I've refactored the code and the tests in eca6ff9 so that now the `NSTemplateTierGenerator` and all its methods are package-protected, and all tests moved in the "whitebox" test file. In the next PR, I'll add a (unique) `CreateOrUpdateResources(...)` which will use this generator, but internally, so it won't be exposed to the outside world. Also, having all the tests in `pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go` makes it easier to maintain, and allows for accessing private fields such as `nstemplatetierGenerator.revisions` when asserting the results with the production data.

Please let me know what you think about it.

alexeykazakov replied:
I was thinking about such tests for the prod templates too but look.. This PR introduces NSTemplateTier generator which can work with any templates. And its functionality covered by the tests (using test templates). So we can be sure that the generator itself works as expected.

So, it's OK if we don't test the production templates themselves here. We should test them of course but it can/should be covered the next PRs when we will pass the prod templates to the generator (we don't yet). Basically it should be covered by `CreateOrUpdateResources` tests.

MatousJobanek commented on test/templates/nstemplatetiers/README.adoc:
```diff
@@ -0,0 +1,5 @@
+= Testing the NSTemplateTier manifest geneator
+
+This package contains yaml files which are used by the link:../pkg/templates/nstemplatetier_generate_test.go[NSTemplateTier generator tests].
```
```suggestion
This package contains yaml files which are used by the link:..pkg/templates/nstemplatetiers/nstemplatetier_generator_test.go[NSTemplateTier generator tests].
```

xcoulon commented on pkg/templates/nstemplatetiers/nstemplatetier_generator_test.go:
```diff
@@ -0,0 +1,190 @@
+				// has no specific values for the `TypeMeta`: the `APIVersion: toolchain.dev.openshift.com/v1alpha1`
+				// and `Kind: NSTemplateTier` should be set by the client using the registered GVK
+				assert.Equal(t, expected.ObjectMeta, actual.ObjectMeta)
+				assert.Equal(t, expected.Spec, actual.Spec)
```
well, I opted for a set of templates in the `test` pkg to actually avoid having to maintain this test when the production templates change. Fro now, I'd rather keep the `newNSTemplateTierFromYAML` func which uses a YAML content, which has the benefit of showing a "visual" representation of the expected manifest. It's not such a big deal IMO if we have to update the test when the namespace templates in `test/templates/nstemplatetiers/` evolve. And the benefit here is that we have a pretty straigthforward and exhaustive comparison (besides the `TypeMeta` part, as I explained in a comment.

MatousJobanek replied:
ok, fair enough :+1:

xcoulon commented:
/test e2e

xcoulon commented:
thanks for the reviews @alexeykazakov and @MatousJobanek 🙌
